coreax/kernels/base.py

```diff
@@ -0,0 +1,503 @@
+            shape stability required by :func:`jax.lax.scan` (used for block iteration).
+
+        :param x: Data matrix, :math:`n \times d`
+        :param y: Data matrix, :math:`m \times d`
+        :param axis: Which axis of the kernel matrix to compute the mean over; a value
+            of `None` computes the mean over both axes
+        :param block_size: Size of matrix blocks to process; a value of :data:`None`
+            sets :math:`B_x = n` and :math:`B_y = m`, effectively disabling the block
+            accumulation; an integer value ``B`` sets :math:`B_y = B_x = B`; a tuple
+            allows different sizes to be specified for ``B_x`` and ``B_y``; to reduce
+            overheads, it is often sensible to select the largest block size that does
+            not exhaust the available memory resources
+        :param unroll: Unrolling parameter for the outer and inner :func:`jax.lax.scan`
+            calls, allows for trade-offs between compilation and runtime cost; consult
+            the JAX docs for further information
+        :return: The (weighted) mean of the kernel matrix :math:`K_{ij}`
+        """
+        operands = x, y
+        inner_unroll, outer_unroll = tree_leaves_repeat(unroll, len(operands))
+        _block_size = tree_leaves_repeat(block_size, len(operands))
+        # Row-mean is the argument reversed column-mean due to symmetry k(x,y) = k(y,x)
+        if axis == 0:
+            operands = operands[::-1]
+            _block_size = _block_size[::-1]
+        (block_x, unpadded_len_x), (block_y, _) = jtu.tree_map(
+            block_data_convert, operands, tuple(_block_size), is_leaf=is_data
+        )
+
+        def block_sum(accumulated_sum: Array, x_block: Data) -> tuple[Array, Array]:
+            """Block reduce/accumulate over ``x``."""
+
+            def slice_sum(accumulated_sum: Array, y_block: Data) -> tuple[Array, Array]:
+                """Block reduce/accumulate over ``y``."""
+                x_, w_x = x_block.data, x_block.weights
+                y_, w_y = y_block.data, y_block.weights
+                column_sum_slice = jnp.dot(self.compute(x_, y_), w_y)
+                accumulated_sum += jnp.dot(w_x, column_sum_slice)
+                return accumulated_sum, column_sum_slice
+
+            accumulated_sum, column_sum_slices = jax.lax.scan(
+                slice_sum, accumulated_sum, block_y, unroll=inner_unroll
+            )
+            return accumulated_sum, jnp.sum(column_sum_slices, axis=0)
+
+        accumulated_sum, column_sum_blocks = jax.lax.scan(
+            block_sum, jnp.asarray(0.0), block_x, unroll=outer_unroll
+        )
+        if axis is None:
+            return accumulated_sum
+        num_rows_padded = block_x.data.shape[0] * block_x.data.shape[1]
+        column_sum_padded = column_sum_blocks.reshape(num_rows_padded, -1).sum(axis=1)
+        return column_sum_padded[:unpadded_len_x]
+
+
+class UniCompositeKernel(ScalarValuedKernel):
```

Comment:
Would it make more sense for this (and `DuoCompositeKernel`, `AdditiveKernel`, `ProductKernel`) to be in `coreax.kernels.scalar_valued` instead? Presumably when vector/function-valued kernels are introduced, we'll want to also have an `AdditiveKernel` (etc.) equivalent for them as well.

Reply:
(If moved, remember to change the changelog comment as well!)

Reply:
If we move these classes then I have come into issues with circular imports as we have to import from the `scalar_valued.py` file which itself imports from `base.py` already. 

I couldn't come up with a way to have `AdditiveKernel` for example live outside the same class as `ScalarValuedKernel` for example

Reply:
You _could_ do 

```python
# __init__.py
from coreax.kernels.base import ScalarValuedKernel
from coreax.kernels.scalar_valued import (AdditiveKernel, ProductKernel, ...)

# base.py
from coreax import kernels
def function(x: kernels.AdditiveKernel): ...

# scalar_valued.py
from coreax import kernels
def other(x: kernels.ScalarValuedKernel): ...
```

and this avoids the circular import issue with some import-logic wizardry.

Alternatively, and I think this is possibly a better solution, you could move `base.py` and `scalar_valued.py` to e.g. `scalar_valued/base.py` and `scalar_valued/kernels.py` respectively, so that in future the other-valued kernels could be in e.g. `function_valued/` and `vector_valued/`. Then when importing from elsewhere, rather than `from coreax.kernels import XYZ`, you instead have to specify and go `from coreax.kernels.scalar_valued import XYZ`. (Then we can separately have `coreax.kernels.scalar_valued.AdditiveKernel` and `coreax.kernels.vector_valued.AdditiveKernel`; if client code needs to use both in the same file they can always just `import coreax.kernels.scalar_valued.AdditiveKernel as ScalarAdditiveKernel`.

Reply:
I am having trouble doing `from coreax import kernels` within the `base.py` and `scalar_valued.py` files, it appears to not recognise the import. Perhaps we can revisit this when I finish the operator-valued kernel implementation to see if this is an issue.

Reply:
Revisiting this with the vector/function-valued kernels seems sensible, so let's leave this until then.